spec/app/domain/authentication/authn-gcp/validate_resource_restrictions_spec.rb

```diff
@@ -0,0 +1,175 @@
+      ),
+      Authentication::AuthnGcp::ResourceRestriction.new(
+        type: gcp_service_account_email_restriction_type,
+        value: gcp_service_account_email_restriction_invalid_value
+      )
+    ]
+  }
+
+  let(:mocked_decoded_token_class_return_valid_values) {double("DecodedToken")}
+  let(:mocked_authenticator_input_return_valid_values) {double("AuthenticatorInput")}
+
+  before(:each) do
+    allow(mocked_decoded_token_class_return_valid_values).to receive(:project_id)
+                                                               .and_return(gcp_project_id_restriction_valid_value)
+    allow(mocked_decoded_token_class_return_valid_values).to receive(:instance_name)
+                                                               .and_return(gcp_instance_name_restriction_valid_value)
+    allow(mocked_decoded_token_class_return_valid_values).to receive(:service_account_id)
+                                                               .and_return(gcp_service_account_id_restriction_valid_value)
+    allow(mocked_decoded_token_class_return_valid_values).to receive(:service_account_email)
+                                                               .and_return(gcp_service_account_email_restriction_valid_value)
+
+    allow(mocked_resource_restrictions_return_4_valid).to receive(:call)
+                                                            .and_return(resource_restrictions_with_4_valid)
+    allow(mocked_resource_restrictions_return_4_valid_and_1_illegal).to receive(:call)
+                                                                          .and_return(resource_restrictions_with_4_valid_and_1_illegal)
+    allow(mocked_resource_restrictions_return_4_invalid_values).to receive(:call)
+                                                                     .and_return(resource_restrictions_with_4_invalid_values)
+
+    allow(mocked_authenticator_input_return_valid_values).to receive(:account)
+                                                               .and_return(valid_account)
+    allow(mocked_authenticator_input_return_valid_values).to receive(:username)
+                                                               .and_return(valid_host)
+    allow(mocked_authenticator_input_return_valid_values).to receive(:credentials)
+                                                               .and_return(mocked_decoded_token_class_return_valid_values)
+
+  end
+
+  #  ____  _   _  ____    ____  ____  ___  ____  ___
+  # (_  _)( )_( )( ___)  (_  _)( ___)/ __)(_  _)/ __)
+  #   )(   ) _ (  )__)     )(   )__) \__ \  )(  \__ \
+  #  (__) (_) (_)(____)   (__) (____)(___/ (__) (___/
+
+  context "A valid resource restrictions configuration" do
+    context "when all resource restrictions values exists in GCP token" do
+      subject do
+        ::Authentication::AuthnGcp::ValidateResourceRestrictions.new(
+          extract_resource_restrictions: mocked_resource_restrictions_return_4_valid
```

Comment:
name not clear, don't couple to number it may change

Reply:
done